### PR TITLE
Sema: Remove TypeLoc from ExplicitCastExpr (via TypeExpr)

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
I'm not sure how I feel about this.

### CODE REVIEW

File: lib/Sema/TypeCheckConstraints.cpp [hunk 1]
```diff
@@ -1916,34 +1916,36 @@ Expr *PreCheckExpression::simplifyTypeConstructionWithLiteralArg(Expr *E) {
...
   if (auto precheckedTy = typeExpr->getInstanceType()) {
-    typeLoc = TypeLoc(typeExpr->getTypeRepr(), precheckedTy);
+    castTy = precheckedTy;
   } else {
-    TypeResolutionOptions options(TypeResolverContext::InExpression);
-    options |= TypeResolutionFlags::AllowUnboundGenerics;
+    const auto options =
+        TypeResolutionOptions(TypeResolverContext::InExpression) |
+        TypeResolutionFlags::AllowUnboundGenerics |
+        TypeResolutionFlags::SilenceErrors;
```
Comment (AnthonyLatsis) [inline lib/Sema/TypeCheckConstraints.cpp:1926]: Note that I silenced a duplicate IUO diagnostic here. Let me know if you think it should be left as a FIXME

File: lib/AST/ASTDumper.cpp [hunk 2]
```diff
@@ -2605,7 +2605,7 @@ class PrintExpr : public ExprVisitor<PrintExpr> {
     if (auto checkedCast = dyn_cast<CheckedCastExpr>(E))
       OS << getCheckedCastKindName(checkedCast->getCastKind()) << ' ';
     OS << "writtenType='";
-    GetTypeOfTypeLoc(E->getCastTypeLoc()).print(OS);
+    GetTypeOfExpr(E->getCastTypeExpr());
```
Comment (CodaFi) [inline lib/AST/ASTDumper.cpp:2608]:
```diff
- GetTypeOfExpr(E->getCastTypeExpr());
+ OS << GetTypeOfExpr(E->getCastTypeExpr()).getString(PO);
```

Comment (AnthonyLatsis) [reply lib/AST/ASTDumper.cpp:2608]: Oops, thank you.

**Comment by AnthonyLatsis** [comment]
@CodaFi Is the plan to completely get rid of TypeLoc, or just across the CS?

**Comment by CodaFi** [comment]
I don't know if we can completely eliminate them in the short or medium term because of SIL, but it would certainly be nice to eventually. They're an attractive nuisance in the AST: 

- They hide double-diagnostics in TypeCheckType
- Different components smuggle type information inside of them
- Some components skip checks when the type is filled in
- They make an invalid state representable: implicit AST nodes can have null type locs

So, at least in the surface language, eliminating them unlocks a lot of simplifications. We should be able to delete `TypeChecker::validateType`, have an easier time refactoring TypeCheckType to not return the null `Type`, and strip CS of its special cases for TypeLocs.

**Comment by AnthonyLatsis** [comment]
Thanks for the break-up, I'd better unpatch the duplicate diagnostic then.

File: lib/Sema/TypeCheckConstraints.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckConstraints.cpp:1926]: This looks reasonable to me, thank you!

**Review by DougGregor — APPROVED** [review]
Nice!

**Review by xedin — COMMENTED** [review]
Looks good! Before we go ahead could we please consider whether it would be possible to use `TypeRepr *` directly instead of `TypeExpr *` for these expressions? I'm a little uneasy that `getType()/getInstanceType()` would lead to confusion, it seems like type could be saved in the expression itself so having a separate expression is unnecessary.

**Comment by CodaFi** [comment]
I don’t know if we can hide the underlying TypeExpr from the constraint system without smuggling the type through something else. All other callers can use the repr or instance type directly, sure.

**Comment by xedin** [comment]
Constraint system is able to save resolved type of `TypeRepr *` for its future reference. Do any other parts need to refer to a type of cast instead of type expression which represents a cast?

**Comment by AnthonyLatsis** [comment]
> Constraint system is able to save resolved type of TypeRepr * for its future reference.

How do we cache it? There is no `setType(TypeRepr *` according to code completion.

**Comment by CodaFi** [comment]
We can’t use a raw TypeRepr node for casts because `is` expressions need the type saved off separately. And if we're just going to store a separate Type alongside a TypeRepr to address that, then we haven’t really won anything from this refactoring. Can we get the best of both worlds by hiding the TypeExpr from all clients except CS?

**Comment by xedin** [comment]
> How do we cache it? There is no setType(TypeRepr * according to code completion.

`setType(ASTNode, ...)` could be used for that. ASTNode has `TypeLoc *` at the moment but it could be changed to `TypeRepr *` which is actually much better and would allow us to remove this special `setType(TypeLoc &, ...)` accessor.

**Comment by xedin** [comment]
> And if we're just going to store a separate Type alongside a TypeRepr to address that, then we haven’t really won anything from this refactoring

But isn't it true that replacing `TypeLoc` with `TypeExpr` does the same?

**Comment by AnthonyLatsis** [comment]
Since replacing TypeLoc with TypeRepr as a ASTNode seems to be more extensive, would it be acceptable to temporarily define a separate cache for `TypeRepr *`?

**Comment by xedin** [comment]
What do you mean by extensive?

**Comment by AnthonyLatsis** [comment]
I mean larger refactoring, but I could be wrong. Do you want me to try that here?

**Comment by xedin** [comment]
My understanding is that we use `TypeExpr` here because we need to store a type of cast somewhere. From API perspective it would be reasonable to use `TypeRepr` directly for these expressions if we didn't need to deal with a type problem, that should close all of the bullet points in your previous comment even more so about smuggling types because there is a way to do that with `TypeExpr` but not with `TypeRepr`.

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

**Comment by AnthonyLatsis** [comment]
@CodaFi Are you okay with having a raw TypeRepr if we lay the type caching on the CS? My only concern is whether we need that type info outside the CS.

**Comment by AnthonyLatsis** [comment]
Oh, I see now, we need the type in `RValueEmitter::visitIsExpr`, and of course we don't have a CS around.

**Comment by xedin** [comment]
> My only concern is whether we need that type info outside the CS

That's exactly my question as well, if we don't than nothing should prevent us from using `TypeRepr` here. Previous refactorings use `TypeExpr` because for those expressions it could be either a `Type` or `TypeRepr`, but here it seems to be a different story.

**Comment by xedin** [comment]
Another question is whether expression itself could be used to indicate a type of a cast instead of it being set on the member...

**Comment by AnthonyLatsis** [comment]
Looks like there's a test that checks the AST dump for a cast expr., hence the failure.

**Comment by AnthonyLatsis** [comment]
@xedin We need a way to pass a `castTo` type for implicit cast expressions, and `TypeLoc` on `IsPattern` is on the list too. Would it make sense to express these not-really-type-expressions as a new kind of simple TypeRepr-storing Pattern of instance type? I think we should try to save type info, so as to at least properly dump such expressions without having to depend on a CS.

**Comment by xedin** [comment]
Yes, that sounds reasonable to me.

**Comment by AnthonyLatsis** [comment]
@CodaFi Do you have an opinion on expressing this with a Pattern?

**Comment by AnthonyLatsis** [comment]
I'm leaving this PR intact in case we still decide to go with TypeExpr. The pattern version is implemented in #32202

File: lib/Sema/TypeCheckPattern.cpp [hunk 3]
```diff
@@ -360,10 +360,11 @@ class ResolvePattern : public ASTVisitor<ResolvePattern,
     auto cast = dyn_cast<CoerceExpr>(E->getElement(1));
     if (!cast)
       return nullptr;
-    
+
+    const auto tyLoc = TypeLoc(cast->getCastTypeExpr()->getTypeRepr());
     Pattern *subPattern = getSubExprPattern(E->getElement(0));
-    return new (Context) IsPattern(cast->getLoc(), cast->getCastTypeLoc(),
-                                   subPattern, CheckedCastKind::Unresolved);
+    return new (Context) IsPattern(cast->getLoc(), tyLoc, subPattern,
```
Comment (CodaFi) [inline lib/Sema/TypeCheckPattern.cpp:366]: Until `IsPattern` is disabused of its TypeLoc, you also have to make sure to fill in the instance type from the cast type here.

**Comment by AnthonyLatsis** [comment]
OK, I've completely hidden the type expression and had CS rely on TypeRepr to do the caching part.

File: lib/Sema/CSApply.cpp [hunk 4]
```diff
@@ -3492,20 +3492,22 @@ namespace {
...
 
-      // Set the type we checked against.
-      expr->getCastTypeLoc().setType(toType);
-      auto fromType = cs.getType(sub);
+      // Simplify and update the type we checked against.
+      auto *const castTypeRepr = expr->getCastTypeRepr();
+
+      const auto fromType = cs.getType(sub);
+      const auto toType = simplifyType(cs.getType(castTypeRepr));
+      expr->setCastType(toType);
```
Comment (AnthonyLatsis) [inline lib/Sema/CSApply.cpp:3503]: @xedin Is there a meaning in updating the cached type as well at this point?

File: include/swift/AST/TypeLoc.h [hunk 5]
```diff
@@ -29,7 +29,7 @@ class TypeRepr;
 
 /// TypeLoc - Provides source location information for a parsed type.
 /// A TypeLoc is stored in AST nodes which use an explicitly written type.
-class alignas(1 << TypeLocAlignInBits) TypeLoc {
+class alignas(8) TypeLoc final {
```
Comment (xedin) [inline include/swift/AST/TypeLoc.h:32]: Could you please keep it as `1 << TypeReprAlignInBits` until `TypeLoc` gets removed?

File: lib/Sema/CSApply.cpp (hunk 4 above)
Comment (xedin) [reply lib/Sema/CSApply.cpp:3503]: Yes, I think that make sense, other re-writer actions might request that type so AST should be kept consistent.

**Review by xedin — APPROVED** [review]
Looks great, thank you, @AnthonyLatsis!

File: include/swift/AST/TypeLoc.h (hunk 5 above)
Comment (AnthonyLatsis) [reply include/swift/AST/TypeLoc.h:32]: Ah, sure thing!

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Review by CodaFi — APPROVED** [review]

**Pull request merged**